meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+              , state : addCommit idCommit person value timeout state }
+              NoProblem
+eval (DPay idCommit person value) state = 
+  if (not (isCurrentCommit idCommit state))
+  then (if (not (isExpiredCommit idCommit state))
+        then Result { outcome : emptyOutcome,
+                      state : state }
+                    CommitNotMade
+        else Result { outcome : emptyOutcome,
+                      state : state}
+                    CommitIsExpired)
+  else (if value > maxValue
+        then case discountAvailableMoneyFromCommit idCommit maxValue state of
+               Just newState -> Result { outcome : addOutcome person maxValue emptyOutcome
+                                       , state : newState }
+                                       NotEnoughMoneyLeftInCommit
+               Nothing -> InconsistentState
+        else case discountAvailableMoneyFromCommit idCommit value state of
+               Just newState -> Result { outcome : addOutcome person value emptyOutcome
+                                       , state : newState }
+                                NoProblem
+               Nothing -> InconsistentState)
+  where maxValue = getAvailableAmountInCommit idCommit state
+
+-- Check whether the person who must sign has signed
+areActionPermissionsValid :: DetachedPrimitive -> S.Set Person -> Boolean
+areActionPermissionsValid (DCommit _ person _ _) sigs = person `S.member` sigs
+areActionPermissionsValid (DPay _ person _) sigs = person `S.member` sigs
+
+areInputPermissionsValid :: Input -> S.Set Person -> Boolean
+areInputPermissionsValid (IChoice cid _) sigs = cid.person `S.member` sigs
+areInputPermissionsValid (IOracle _ _ _) _ = true -- Implementation ToDo: need to check signature 
+
+-- Check whether a commit or payment has negative value
+isTransactionNegative :: DetachedPrimitive -> Boolean
+isTransactionNegative (DCommit _ _ val _) = val < (fromInt 0)
+isTransactionNegative (DPay _ _ val) = val < (fromInt 0)
+
+data ErrorResult = InvalidInput
+                 | NoValidSignature
+                 | NegativeTransaction
+                 | AmbiguousId
+                 | InternalError -- This should not happen
+derive instance eqErrorResult :: Eq ErrorResult  
+derive instance ordErrorResult :: Ord ErrorResult  
+
+data ApplicationResult a = SuccessfullyApplied a DynamicProblem
+                         | CouldNotApply ErrorResult
+
+-- High level wrapper that calls the appropriate update function on contract and state.
+-- Does not take care of reducing, that must be done before and after applyAnyInput.
+applyAnyInput :: AnyInput -> S.Set Person -> S.Set IdInput -> BlockNumber -> State -> Contract -> ApplicationResult {outcome :: TransactionOutcomes,
+                        state :: State,
+                        contract :: Contract}
+applyAnyInput anyInput sigs neededInputs blockNum state contract = 
+  case addAnyInput blockNum anyInput neededInputs state of
```

Comment:
You can consider:

```
case anyInput, addAnyInput blockNum anyInput neededInputs state of
  Just updatedState, Input input -> ...
```

ie, you can combine the case instead of nesting them, though the PS syntax is slightly different.

Reply:
Ah! I didn't know you could do that. Will fix
Thank you